src/Std/Sat/AIG/RefStream.lean

```diff
@@ -0,0 +1,188 @@
+/-
+Copyright (c) 2024 Lean FRO, LLC. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Henrik Böving
+-/
+import Std.Sat.AIG.LawfulOperator
+import Std.Sat.AIG.CachedGatesLemmas
+
+namespace Std
+namespace Sat
+
+namespace AIG
+
+variable {α : Type} [Hashable α] [DecidableEq α] {aig : AIG α}
+
+namespace RefStream
+
+def empty : RefStream aig 0 where
+  refs := #[]
+  hlen := by simp
+  hrefs := by intros; contradiction
+
+@[inline]
+def cast' {aig1 aig2 : AIG α} (s : RefStream aig1 len)
+    (h :
+      (∀ {i : Nat} (h : i < len), s.refs[i]'(by have := s.hlen; omega) < aig1.decls.size)
+        → ∀ {i : Nat} (h : i < len), s.refs[i]'(by have := s.hlen; omega) < aig2.decls.size) :
+    RefStream aig2 len :=
+  { s with
+      hrefs := by
+        intros
+        apply h
+        · intros
+          apply s.hrefs
+          assumption
+        · assumption
+  }
+
+@[inline]
+def cast {aig1 aig2 : AIG α} (s : RefStream aig1 len) (h : aig1.decls.size ≤ aig2.decls.size) :
+    RefStream aig2 len :=
+  s.cast' <| by
+    intro hall i hi
+    specialize hall hi
+    omega
+
+@[inline]
+def get (s : RefStream aig len) (idx : Nat) (hidx : idx < len) : Ref aig :=
+  let ⟨refs, hlen, hrefs⟩ := s
+  let ref := refs[idx]'(by rw [hlen]; assumption)
+  ⟨ref, by apply hrefs; assumption⟩
+
+@[inline]
+def push (s : RefStream aig len) (ref : AIG.Ref aig) : RefStream aig (len + 1) :=
+  let ⟨refs, hlen, hrefs⟩ := s
+  ⟨
+    refs.push ref.gate,
+    by simp [hlen],
+    by
+      intro i hi
+      simp only [Array.get_push]
+      split
+      . apply hrefs
+        omega
+      . apply AIG.Ref.hgate
+  ⟩
+
+@[simp]
+theorem get_push_ref_eq (s : RefStream aig len) (ref : AIG.Ref aig) :
+    (s.push ref).get len (by omega) = ref := by
+  have := s.hlen
+  simp [get, push, ← this]
+
+-- This variant exists because it is sometimes hard to rewrite properly with DTT
+theorem get_push_ref_eq' (s : RefStream aig len) (ref : AIG.Ref aig) (idx : Nat)
```

Comment:
```suggestion
-- This variant exists because it is sometimes hard to rewrite properly with DTT.
theorem get_push_ref_eq' (s : RefStream aig len) (ref : AIG.Ref aig) (idx : Nat)
```